Index: includes/UserMailer.php
===================================================================
--- includes/UserMailer.php	(revision 58851)
+++ includes/UserMailer.php	(working copy)
@@ -106,8 +106,10 @@
 	 * @return mixed True on success, a WikiError object on failure.
 	 */
 	static function send( $to, $from, $subject, $body, $replyto=null, $contentType=null ) {
-		global $wgSMTP, $wgOutputEncoding, $wgErrorString, $wgEnotifImpersonal;
+		global $wgSMTP, $wgOutputEncoding, $wgErrorString, $wgEnotifImpersonal, $wgEmailContentType;
 		global $wgEnotifMaxRecips;
+		if (!$wgEmailContentType)
+			$wgEmailContentType = 'text/plain';
 
 		if ( is_array( $to ) ) {
 			wfDebug( __METHOD__.': sending mail to ' . implode( ',', $to ) . "\n" );
@@ -145,7 +147,7 @@
 			$headers['Date'] = date( 'r' );
 			$headers['MIME-Version'] = '1.0';
 			$headers['Content-type'] = (is_null($contentType) ?
-					'text/plain; charset='.$wgOutputEncoding : $contentType);
+					$wgEmailContentType.'; charset='.$wgOutputEncoding : $contentType);
 			$headers['Content-transfer-encoding'] = '8bit';
 			$headers['Message-ID'] = "<$msgid@" . $wgSMTP['IDHost'] . '>'; // FIXME
 			$headers['X-Mailer'] = 'MediaWiki mailer';
@@ -369,6 +371,10 @@
 					// Already sent an email to this person
 					$userCondition .= ' AND wl_user != ' . intval( $userTalkId );
 				}
+
+				$userCondition .= ' AND wl_notificationtimestamp IS NULL';
+				wfRunHooks('EnotifUserCondition', array(&$this, &$userCondition));
+
 				$dbr = wfGetDB( DB_SLAVE );
 
 				list( $user ) = $dbr->tableNamesN( 'user' );
@@ -380,7 +386,6 @@
 						'wl_title' => $title->getDBkey(),
 						'wl_namespace' => $title->getNamespace(),
 						$userCondition,
-						'wl_notificationtimestamp IS NULL',
 					), __METHOD__ );
 				$userArray = UserArray::newFromResult( $res );
 
@@ -476,6 +481,7 @@
 		$keys['$PAGEMINOREDIT']      = $medit;
 		$keys['$PAGESUMMARY']        = $summary;
 
+		wfRunHooks('EnotifComposeCommonMailtext', array(&$this, &$keys));
 		$subject = strtr( $subject, $keys );
 
 		# Reveal the page editor's address as REPLY-TO address only if
@@ -580,6 +586,8 @@
 			$wgLang->timeanddate( $this->timestamp, true, false, $timecorrection ),
 			$body);
 
+		wfRunHooks('EnotifPersonalizeMailtext', array(&$this, &$watchingUser, &$body));
+
 		return UserMailer::send($to, $this->from, $this->subject, $body, $this->replyto);
 	}
 
